### ISSUE TITLE
How to extract the scales

### ISSUE THREAD

**Issue by hailo231** [NVIDIA/TransformerEngine#2071]
Hi All , I am doing the Inference only to benchmark a Dense layer how fp8 is doing , i have a pretrained model and extracted all params of dense layer .So if i want to know the scale which are being used here how to extract it **?**
 (i am not training ) i just want to do inference . Will i be able to get scales by this method?
```
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import transformer_engine.pytorch as te
from transformer_engine.common import recipe
batch_size = 1
seq_len = 7
in_features = 3072
out_features = 768

#Load data
inp_np = np.loadtxt("x1.txt", dtype=np.float32).reshape(batch_size, seq_len, in_features)
w_np = np.loadtxt("w1.txt", dtype=np.float32).reshape(in_features, out_features).T
b_np = np.loadtxt("bias.txt", dtype=np.float32).reshape(out_features)
inp_flat = torch.tensor(inp_np.reshape(-1, in_features), device="cuda")
w = torch.tensor(w_np, device="cuda")
b = torch.tensor(b_np, device="cuda")

# Padding for FP8 requirements 
pad_rows = (8 - inp_flat.shape[0] % 8) % 8
pad_cols = (16 - inp_flat.shape[1] % 16) % 16
if pad_rows > 0 or pad_cols > 0:
    inp_padded = F.pad(inp_flat, (0, pad_cols, 0, pad_rows))
else:
    inp_padded = inp_flat

# nn.Linear (FP32 baseline) 
linear_pt = nn.Linear(in_features, out_features, bias=True).cuda()
with torch.no_grad():
    linear_pt.weight.copy_(w)
    linear_pt.bias.copy_(b)

out_pt = linear_pt(inp_padded)[:inp_flat.shape[0], :out_features].reshape(batch_size, seq_len, out_features)
out_pt_flattened = out_pt.ravel()
print("PyTorch Linear output shape:", out_pt.shape)
print(out_pt_flattened[:100])

# Transformer Engine Linear (FP8 autocast)
linear_te = te.Linear(in_features, out_features, bias=True).cuda()
with torch.no_grad():
    linear_te.weight.copy_(w)
    linear_te.bias.copy_(b)

fp8_recipe = recipe.DelayedScaling(margin=0, interval=1, fp8_format=recipe.Format.E4M3)
with te.fp8_autocast(enabled=True, fp8_recipe=fp8_recipe):
    out_te = linear_te(inp_padded)[:inp_flat.shape[0], :out_features].reshape(batch_size, seq_len, out_features)


out_te_flattened = out_te.ravel()
print("Transformer Engine Linear output shape:", out_te.shape)
print(out_te_flattened[:100])
mse = F.mse_loss(out_pt, out_te).item()
mae = F.l1_loss(out_pt, out_te).item()
rmse = torch.sqrt(F.mse_loss(out_pt, out_te)).item()
max_diff = torch.max(torch.abs(out_pt - out_te)).item()

print(f"MSE:      {mse:.8f}")
print(f"RMSE:     {rmse:.8f}")
print(f"MAE:      {mae:.8f}")
print(f"Max diff: {max_diff:.8f}")
```

**Comment by zhongbozhu** [comment]
You can find the `base` classes in the folder `transformer_engine/pytorch/tensor/_internal`. For example, in the case of FP8 delayed scaling, here is the data structure that you can access and print out: https://github.com/NVIDIA/TransformerEngine/blob/757fd1cfd7ef09c664df5dc854e1051665e3317f/transformer_engine/pytorch/tensor/_internal/float8_tensor_base.py#L55

Please note that when the scales living in the quantized tensor is always `inverse of the scale used in cast`. Scales don't need to exist in the quantized tensor, because when you apply the scaling factor in the cast, GEMM only need its inverse value to output a high precision value.

**Comment by hailo231** [comment]
Thanks for the response , wanted to understand is this the correct depiction of dataflow of te.Linear and  so the scales i am getting are of which fp8(e4m3) since  gemm expects fp8 format or fp16 casting ? or i am missing something !! 
 
<img width="771" height="806" alt="Image" src="https://github.com/user-attachments/assets/b937330a-2c32-48f1-aa5a-da9dac453dad" />

**Comment by ptrendx** [comment]
The first picture is mostly correct, the only small thing I would say there is that the scale is 448/amax only in the simplest case in delayed scaling (or when we do current scaling) and we actually use max(amax) over a window of iterations.

The second picture is not correct:
- we only cast to FP16 when AMP (automatic mixed precision) is enabled in pytorch
- we do not do fake quantization, when FP8 is disabled we do not use it at all

As for the extracting the scales - the easiest way would be to take the quantizers used by the Linear layer and see what the scales are, See https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/module/linear.py#L1467 - this will give you all quantizers used in this module, then you can just grab the values of the scales: https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/tensor/float8_tensor.py#L48

One thing to also note is that in order for TE to calculate those scales, you need to pass some data through the Linear layer first - just run a few representative inputs through it. You do need to be in the training mode for this to work though (you do not need to run backward) - if you set eval mode then TE will treat the scales it has (default ones being 1) as static scales and not update them.

Does that help?
